src/qililab/calibration/calibration_controller.py

```diff
@@ -0,0 +1,527 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=anomalous-backslash-in-string
+"""Automatic-calibration Controller module, which works with notebooks as nodes."""
+from datetime import datetime, timedelta
+
+import networkx as nx
+
+from qililab.calibration.calibration_node import CalibrationNode
+from qililab.config import logger
+from qililab.data_management import build_platform, save_platform
+from qililab.platform.platform import Platform
+
+
+class CalibrationController:
+    """Controls the automatic calibration sequence.
+
+    **Usage:**
+        - To calibrate the full graph, use the ``run_automatic_calibration()`` method.
+        - To target and ensure a node works, use the ``maintain(node)`` method.
+
+    |
+
+    **Graph Structure:**
+
+    In the graph, directions should be given by `nodes` pointing to their next `dependents` (natural time flow for calibration).
+    This defines our `starts` and `ends` of the calibration:
+
+    - `starts`: `Roots` (``in_degree=0``, no arrows pointing in, no `dependencies`) of the graph, where all arrows leave the node.
+
+    - `ends`: `Leaves` (``out_degree=0``, no arrows pointing out, no `dependants`) of the graph, where all arrows enter the node.
+
+    .. code-block:: python
+
+        #                     /--> 2 -->\\
+        #                    /     |     \\
+        #    (start, root)  0      |      3 --> 4 (end, leave)
+        #                    \     v     /
+        #                     \--> 1 -->/
+
+    |
+
+    **Calibration Workflow:**
+
+    The calibration process is structured into three levels of methods:
+
+        1. **Highest Level Method**: The ``run_automatic_calibration()`` method finds all the end nodes of the graph (`leaves`, those without further `dependents`) and runs ``maintain()`` on them.
+
+        2. **Mid-Level Methods**: ``maintain()`` and ``diagnose()``.
+            - ``maintain(node)`` starts from the `roots` that ``node`` depends on, and moves forwards (`dependency -> dependant`) until ``node``, checking the last time executions and data at each step. If a problem (``bad_data``) is found, it calls ``diagnose()`` to solve it.
+            - ``diagnose(node)`` does more strict checks, fixing inaccuracies in the system's state to allow ``maintain()`` to continue. It works in reverse, starting from the problematic (``bad_data``) node, it goes back (`dependency <- dependant`) until it finds the origin of the problem.
+
+        3. **Low-Level Methods**: ``check_state()``, ``check_data()``, and ``calibrate()`` are the methods you would be calling during this process to interact with the ``nodes``.
```

Comment:
I would remove the "tab" here, because sphinx understands this as a quote:
![Uploading image.png…]()

Reply:
I'll look into it, but I have already tried lots of ways to do it, and there is a problem with lots of them!

Like one fucks up the 1, 2, 3. numeration, another puts everything in a line as if it weren't a new line, but yes, I'll try again!

Reply:
Changed!